ambry-network/src/main/java/com.github.ambry.network/BlockingChannelConnectionPool.java

```diff
@@ -223,6 +223,15 @@ public void destroyBlockingChannel(BlockingChannel blockingChannel) {
       synchronized (lock) {
         // decrement the number of connections to the host and port. we were not able to maintain the count
         numberOfConnections.decrementAndGet();
+        // at this point we are good to clean up the available connections since re-creation failed
+        do {
```

Comment:
You need to take a write lock on `rwLock` before you do this (like `cleanup()`). 

Otherwise there might be a correctness issue here (i.e. you might not close all available connections) because this introduces a race condition with `releaseBlockingChannel()`.